File: Lab_2/Lab_2.1/FloatRange.cpp
#include "FloatRange.h"
#include <sstream>

FloatRange::FloatRange() : first(0), second(0) {}

FloatRange::FloatRange(double f, double s) : first(f), second(s) {}

FloatRange::FloatRange(const FloatRange& other) : first(other.first), second(other.second) {}

FloatRange::~FloatRange() {}

void FloatRange::SetFirst(double f) {
	first = f;
}

void FloatRange::SetSecond(double s) {
	second = s;
}

double FloatRange::GetFirst() const {
	return first;
}

double FloatRange::GetSecond() const {
	return second;
}

FloatRange& FloatRange::operator=(const FloatRange& other) {
	if (this != &other) {
		first = other.first;
		second = other.second;
	}
	return *this;
}

FloatRange::operator string() const {
	stringstream ss;
	ss << "First value: " << first << endl;
	ss << "Second value: " << second << endl;
	return ss.str();
}

FloatRange operator+(const FloatRange& x, const FloatRange& y) {
	return FloatRange(x.first + y.first, x.second + y.second);
}

FloatRange operator-(const FloatRange& x, const FloatRange& y) {
	return FloatRange(x.first - y.first, x.second - y.second);
}

FloatRange operator*(const FloatRange& x, const FloatRange& y) {
	return FloatRange(x.first * y.first, x.second * y.second);
}

FloatRange operator/(const FloatRange& x, const FloatRange& y) {
	if (y.first == 0 || y.second == 0) {
		cout << "Division by zero!" << endl;
		return FloatRange();
	}
	return FloatRange(x.first / y.first, x.second / y.second);
}

FloatRange operator^(const FloatRange& x, const unsigned n) {
	FloatRange result(1, 1);
	for (unsigned i = 0; i < n; ++i) {
		result.first *= x.first;
		result.second *= x.second;
	}
	return result;
}

ostream& operator<<(ostream& out, const FloatRange& range) {
	out << string(range);
	return out;
}

istream& operator>>(istream& in, FloatRange& range) {
	cout << "Enter first value: ";
	in >> range.first;
	cout << "Enter second value: ";
	in >> range.second;
	return in;
}

FloatRange& FloatRange::operator++() {
	first++;
	return *this;
}

FloatRange& FloatRange::operator--() {
	first--;
	return *this;
}

FloatRange FloatRange::operator++(int) {
	FloatRange temp(*this);
	second++;
	return temp;
}

FloatRange FloatRange::operator--(int) {
	FloatRange temp(*this);
	second--;
	return temp;
}















//#include "FloatRange.h"
//FloatRange& operator=(const FloatRange& other) {
//	if (this != &other) {
//		first = other.first;
//		second = other.second;
//	}
//	return *this;
//}
//
//// Операція виводу
//friend std::ostream& operator<<(std::ostream& out, const FloatRange& range) {
//	out << "First value: " << range.first << std::endl;
//	out << "Second value: " << range.second << std::endl;
//	return out;
//}
//
//// Операція вводу
//friend std::istream& operator>>(std::istream& in, FloatRange& range) {
//	std::cout << "Enter first value: ";
//	in >> range.first;
//	std::cout << "Enter second value: ";
//	in >> range.second;
//	return in;
//}
//
//// Операція приведення типу - перетворення у літерний рядок
//operator std::string() const {
//	std::stringstream ss;
//	ss << "First value: " << first << std::endl;
//	ss << "Second value: " << second << std::endl;
//	return ss.str();
//}
//
//// Операція інкременту (префіксна)
//FloatRange& operator++() {
//	++first;
//	return *this;
//}
//
//// Операція декременту (префіксна)
//FloatRange& operator--() {
//	--first;
//	return *this;
//}
//
//// Операція інкременту (постфіксна)
//FloatRange operator++(int) {
//	FloatRange temp(*this);
//	++first;
//	return temp;
//}
//
//// Операція декременту (постфіксна)
//FloatRange operator--(int) {
//	FloatRange temp(*this);
//	--first;
//	return temp;
//}

File: Lab_2/Lab_2.1/FloatRange.cs
using System;

namespace Lab_1._1;

public class FloatRange
{
    private double First { get; set; }
    private double Second { get; set; }

    public FloatRange(double first, double second)
    {
        if (first < second)
        {
            First = first;
            Second = second;
        }
        else
        {
            Console.WriteLine("Error, first value is not less than second value");
            // Setting default values if the condition is not met
            First = 0;
            Second = 0;
        }
    }

    public FloatRange()
    {
        // Default constructor without arguments
        First = 0;
        Second = 0;
    }

    // Copy constructor
    public FloatRange(FloatRange other)
    {
        First = other.First;
        Second = other.Second;
    }

    public void Read()
    {
        double firstValue;
        double secondValue;
        do
        {
            Console.Write("Enter first value: ");
            firstValue = Convert.ToDouble(Console.ReadLine());

            Console.Write("Enter second value: ");
            secondValue = Convert.ToDouble(Console.ReadLine());
        }
        while (firstValue >= secondValue);

        First = firstValue;
        Second = secondValue;
    }

    public void Display()
    {
        Console.WriteLine($"First value: {First}");
        Console.WriteLine($"Second value: {Second}");
    }

    public bool RangeCheck(double x)
    {
        return (First <= x && x <= Second);
    }


    public static void Read(FloatRange floatRange)
    {
        floatRange.Read();
    }

    public static void Display(FloatRange floatRange)
    {
        floatRange.Display();
    }
    public static FloatRange operator +(FloatRange x, FloatRange y)
    {
        return new FloatRange(x.First + y.Second, x.First + y.Second);
    }
    public static FloatRange operator -(FloatRange x, FloatRange y)
    {
        return new FloatRange(x.First - y.Second, x.First - y.Second);
    }

    public static FloatRange operator *(FloatRange x, FloatRange y)
    {
        return new FloatRange(x.First * x.Second, y.First * y.Second);
    }

    public static FloatRange operator /(FloatRange x, FloatRange y)
    {
        if (y.First == 0 || y.Second == 0)
        {
            Console.WriteLine("Division by zero!");
            return new FloatRange();
        }
        return new FloatRange(x.First / y.First, x.Second / y.Second);
    }

    public static FloatRange operator ^(FloatRange x, int n)
    {
        FloatRange result = new FloatRange(1, 1);
        for (int i = 0; i < n; ++i)
        {
            result.First *= x.First;
            result.Second *= x.Second;
        }
        return result;
    }
    // Method for postfix increment (range++)
    public static FloatRange operator ++(FloatRange range)
    {
        range.Second++;
        return range;
    }
    // Method for postfix decrement (range--)
    public static FloatRange operator --(FloatRange range)
    {
        range.Second--;
        return range;
    }

    // Prefix increment operator (range++)
    public FloatRange PrefixIncrement()
    {
        FloatRange temp = new FloatRange(First, Second);
        First++;
        return temp;
    }
    // Prefix decrement operator (range--)
    public FloatRange PrefixDecrement()
    {
        FloatRange temp = new FloatRange(First, Second);
        First--;
        return temp;
    }

    // Conversion to string
    public static implicit operator string(FloatRange floatRange)
    {
        return $"First value: {floatRange.First}, Second value: {floatRange.Second}";
    }
}


